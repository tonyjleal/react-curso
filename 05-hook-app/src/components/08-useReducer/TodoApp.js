import React, { useReducer } from 'react';
import './style.css';
import { todoReducer } from './todoReducer';


export const TodoApp = () => {

 const init = () => {
  return [{
    id: new Date().getTime(),
    desc: 'Aprender React', 
    done: false
  }];
 }
 
 const [ todos, dispatch ]  = useReducer(todoReducer, [], init);


  const handleSubmit = (e) => {
    e.preventDefault();

    const newTodo = {
      id: new Date().getTime(),
      desc: 'New Task',
      done: false
    }
    
    const action = {
      type: 'add',
      payload: newTodo
    }

    dispatch(action);

  }


  return(
     <div>

        <h1> TODO APP ( { todos.length })</h1>
        <hr />
        <div  className='row'
        >
          <div className='col-7'>
            <ul className='list-group list-group-flush'>
              {
                todos.map((todo, i) => {
                  return <li
                          key={todo.id}
                          className='list-group-item'>
                          <p className='text-center'>{ i +1 }. { todo.desc }</p> 
                            <button className='btn btn-danger'>
                              Borrar
                            </button>
                          </li>
                })
              }
            </ul>
          </div>
          <div className='col-5'>
            <h4>Agregar TODO</h4>
            <hr />

            <form onSubmit={handleSubmit}>
              <input 
              className='form-control'
              type="text" 
              name="description"
              placeholder='Aprender...'
              autoComplete='off' />

              <button type='submit' className='btn btn-outline-primary col-12 mt-1'>
                Agregar
              </button>
            </form>
          </div>

        </div>
    </div>
    );
};
