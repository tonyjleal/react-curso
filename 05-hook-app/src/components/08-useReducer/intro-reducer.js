
const initialState = [{
    id: 1,
    todo: 'Buy milk',
    done: false,
}]


const todoReducer = ( state = initialState, action ) => {

    if (action?.type === 'add') {
        return [...state, action.payload];
    }

    return state;
}

let todos = todoReducer();

const newTodo = {
    id: 2,
    todo: 'Buy cookies',
    done: false,
}

const action = {
    type: 'add',
    payload: newTodo
}

todos = todoReducer(todos, action)

console.log(todos);