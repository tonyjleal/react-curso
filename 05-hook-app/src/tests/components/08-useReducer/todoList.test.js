import {shallow} from 'enzyme'
import { TodoList } from '../../../components/08-useReducer/TodoList'
import { demoTodos } from '../../fixtures/demoTodos';

describe('Test component <TodoList />', () => { 

    const handleDelete = jest.fn();
    const handleToggle = jest.fn();
    const wrapper = shallow(
        <TodoList 
            todos={demoTodos}
            handleDelete={handleDelete}
            handleToggle={handleToggle} 
        />
    )

    test('should show component success', () => { 

        expect(wrapper).toMatchSnapshot();

     })

    test('should have two <TodoListItem />', () => { 

        const elements = wrapper.find('TodoListItem')

        expect(elements.length).toBe(demoTodos.length)

     })

    test('should have handleDelete function', () => { 
          
        const functionDelete = wrapper.find('TodoListItem').at(0).prop('handleDelete');
        
        console.log(functionDelete)

        expect(functionDelete).toEqual(expect.any(Function))

     })
    
 })