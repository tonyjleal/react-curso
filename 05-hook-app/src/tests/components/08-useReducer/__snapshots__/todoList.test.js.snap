// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test component <TodoList /> should show component success 1`] = `
<ul
  className="list-group list-group-flush"
>
  <TodoListItem
    handleDelete={[MockFunction]}
    handleToggle={[MockFunction]}
    index={0}
    key="1"
    todo={
      Object {
        "desc": "Aprender React",
        "done": false,
        "id": 1,
      }
    }
  />
  <TodoListItem
    handleDelete={[MockFunction]}
    handleToggle={[MockFunction]}
    index={1}
    key="2"
    todo={
      Object {
        "desc": "Aprender Mongo",
        "done": false,
        "id": 2,
      }
    }
  />
</ul>
`;
